pwa/src/templates/templateParts/mappingForm/MappingFormTemplate.tsx

```diff
@@ -26,9 +28,18 @@ export const MappingFormTemplate: React.FC<MappingFormTemplateProps> = ({ mappin
   const { t } = useTranslation();
   const { setIsLoading, isLoading } = useIsLoadingContext();
 
-  const [_mapping, setMapping] = React.useState<any[]>([]);
-  const [unset, setUnset] = React.useState<any[]>([]);
-  const [cast, setCast] = React.useState<any[]>([]);
+  const [mappingKeyValue, setMappingKeyValue] = React.useState<any[]>([]);
+  const [mappingCodeEditor, setMappingCodeEditor] = React.useState<string>("");
+  const [editMappingInEditor, setEditMappingInEditor] = React.useState<boolean>(false);
+
+  const [unsetKeyValue, setUnsetKeyValue] = React.useState<any[]>([]);
+  const [unsetCodeEditor, setUnsetCodeEditor] = React.useState<string>("");
+  const [editUnsetInEditor, setEditUnsetInEditor] = React.useState<boolean>(false);
+
+  const [castKeyValue, setCastKeyValue] = React.useState<any[]>([]);
+  const [castCodeEditor, setCastCodeEditor] = React.useState<string>("");
+  const [editCastInEditor, setEditCastInEditor] = React.useState<boolean>(false);
+
   const [isOpenMapping, setIsOpenMapping] = React.useState<boolean>(!mapping);
   const [isOpenUnset, setIsOpenUnset] = React.useState<boolean>(!mapping);
   const [isOpenCast, setIsOpenCast] = React.useState<boolean>(!mapping);
```

Comment:
(See other comment about removing `mappingCodeEditor`, `unsetCodeEditor` and `castCodeEditor`).

It might be possible to simplify this logic. Basically, we want to keep track of three things (I think):

1. Which one (of three) types of mappings are open;
2. Which one (of three) types is being edited as an JSON;
3. The value of each one.

Currently, we have 12 states to keep track of this. Not necessarily an issue, however, it might be worthwhile to check if we can reduce that number.

Possible (**dummy!**) solution:

```
type TMappingTypeState = {
  open: boolean;
  editAsJSON: boolean;
  value: string;
}

interface IMappingState {
  mapping: TMappingTypeState,
  unset: TMappingTypeState,
  cast: TMappingTypeState,
}

const [mappingsState, setMappingsState] = React.useState<IMappingState>({ ...default state });

// I think we could run all further logic from this one state, making it easier to be managed.
```